text/dependencies/rfcs/0000-dependency-management-phase-one.md

```diff
@@ -0,0 +1,416 @@
+
+2. **Compilation**. Almost every dependency is compiled from source, or has
+   some kind of processing for simple installation in the buildpack build
+   process. Dependencies are compiled against Ubuntu 18.04 which means the
+   dependencies are not necessarily compatible with other Stacks. This means
+   the related buildpacks are locked to running on Ubuntu 18.04 based stacks
+   only. Since dependencies are compiled, they are stored in an S3 bucket which
+   is costly for an ever-expanding set of dependencies and versions.
+
+3. **Usability**. The current system is difficult to run outside of our
+   automation, especially the dependency compilation code. When compilation
+   code has to change for a new dependency version, it's very challenging to
+   use the code locally to test changes. Similarly, the workflows that run in
+   the dep-server are not intuitve to use locally. All of the dependency
+   management process should ideally able to run locally with ease, in the
+   event of Github Actions outages.
+
+### Overarching Plan
+
+The various parts of the dependency management process can be distilled into a
+few high-level jobs: (1) new version retrieval, (2) metadata gathering, (3)
+compilation, (4) dependency testing, and a very important new addition is (5),
+performing these steps such that dependencies are compatible with different
+stacks/operating systems/architectures. All 5 of these steps require different
+solutions depending on the dependency, which is why they will become the
+responsibility of the **buildpack author/maintainer**.
+
+Buildpack authors will be responsible for creating code to perform each of
+these steps, which will live in a new buildpack directory called
+`dependency`. It will be the job of generalized workflows, actions, and tooling
+to orchestrate these steps in an automated fashion, to keep the
+`buildpack.toml` up-to-date with new dependency versions.
+
+In order to run the steps in a generalized workflow, each step will need to
+conform to a standardized set of inputs and outputs. Each can be written in any
+language, but should have a simple way to be executed via a Makefile so it can
+be called in a consistent manner. A `Makefile` will be created with targets for
+each step, with each step being a bash script in this example:
+
+`<buildpack>/dependency/Makefile`:
+```
+retrieve:
+	@./retrieval/retrieve.sh $(buildpackTomlPath) $(output)
+compile:
+	@./compilation/compile.sh $(version) $(outputDir) $(target)
+test:
+	@./test/test.sh $(tarballName) $(version)
+```
+
+#### 1. Version Retrieval (`make retrieve`)
+  * Description: Discovers dependency versions newer than `buildpack.toml`
+    versions (within constraints) and returns each with a set of generated
+    metadata in the form of a `metadata.json` file.
+  * Inputs:
+    * `buildpackTomlPath`  - the path to the  `buildpack.toml` file
+    * `output` - the path to a file into which an output metadata JSON will be
+      written
+  * Outputs:
```

Comment:
Once it's solidified, a sample JSON output that represents the output schema would be helpful here.

Reply:
added

Reply:
There are a small handful of JSON files used to pass information around through the workflow in a structured way.

I suspect we could add golang struct representations of these in `libdependencies` that could be the ongoing "canonical" representation of the JSON. Buildpack authors can choose any language to produce or ingest the JSON but should realize that the workflow will probably rely on the representation from `libdependencies`.

Reply:
Depending on Java-side buy in, we may loosen this schema a bit. TBD.

Reply:
Added one for the sake of an example though anyways.

Reply:
Is this schema more solidified now? Are there optional vs. required fields?

Reply:
@dmikusa-pivotal  [cited concerns](https://github.com/paketo-buildpacks/rfcs/pull/204#discussion_r915930540) around differences in metadata approaches, but I think that this schema should be solidified in order for all of the workflows to leverage the same "assembly" step to put stuff into the `buildpack.toml`. I'll add clarity to this in the RFC.

Reply:
I am slightly confused as to whether there will be one `metadata.json` file for each version (or each constraint?) or whether there will be a single `metadata.json` file with an array of objects described below. Either could work, but I'd appreciate some more clarity.

Reply:
It's a single `metadata.json` file! I added a little bit of language clarity to some of the sections